### ISSUE TITLE
Ability to pass headers from ESP to grpc backend

### ISSUE THREAD

**Issue by sgammon** [cloudendpoints/esp#323]
I don't see a way, documented or otherwise, to pass headers to gRPC similar to `proxy_set_header` in Nginx. I figured, perhaps, that all headers were transcoded to metadata, but that doesn't seem to be the case.

Is it possible to add a directive where we can set headers in the proxy, and see them in gRPC? For instance, if behavior in gRPC is based on source IP or some other transport value that isn't available without some headers?

I would be happy to contribute however I can. I'm running gRPC and ESP on Kubernetes in grpc-java.

**Comment by qiwzhang** [comment]
Hi @sgammon My understanding is: ESP should pass all client headers into gRPC metadata for  both gRPC client and HTTP client.  The code is here:  https://github.com/cloudendpoints/esp/blob/master/src/nginx/grpc.cc#L234

If that is NOT the case,  please help us debug it.

**Comment by lizan** [comment]
@qiwzhang I think what @sgammon try to do is to insert a header by ESP, which is not sent by client.

@sgammon I don't have time to work on this very soon. If you can contribute I'm happy to review though.

Some notes on implementation:
- Add a member to the location config `ngx_esp_loc_conf_t`, and parse it from nginx.conf
- The `ngx_esp_loc_conf_t` is extracted for gRPC here: https://github.com/cloudendpoints/esp/blob/master/src/nginx/grpc.cc#L154 
- Add the headers to https://github.com/cloudendpoints/esp/blob/master/src/nginx/grpc.cc#L174

**Comment by emicklei** [comment]
@sgammon can you confirm the statement of @qiwzhang saying "ESP should pass all client headers into gRPC metadata for both gRPC client and HTTP client."  
We have an issue passing our custom headers such "ctx_locale" ; this is absent in our gRPC server.

**Comment by emicklei** [comment]
hi, we found this issue. 
Because the custom headers included and underscore "_" in their name was the reason it was not passed.

**Comment by liam-j-bennett** [comment]
Has there been any progress on this? I'm looking for a way to send on the real IP of the remote client to gRPC using esp (similar to ```proxy_set_header X-Real-IP $remote_addr;``` for non-grpc). Wouldn't mind contributing, but can't say I'm very familar with nginx.

**Comment by qiwzhang** [comment]
Ideally,  ESP should support "proxy_set_header" flag for grpc backend to set any Nginx variables into grpc metadata. 

Unfortunately, There is no much progress in this feature.

**Comment by lucemia** [comment]
https://github.com/cloudendpoints/esp/pull/384#discussion_r183524675

Just need to pass in a --underscores_in_headers flag. 😀

**Comment by lucemia** [comment]
```
kubectl edit deployment gnlp-0-1-5-3  

...
spec:
  progressDeadlineSeconds: 600
  replicas: 1
  revisionHistoryLimit: 10
  selector:
    matchLabels:
      app: gnlp-0-1-5-3
  strategy:
    rollingUpdate:
      maxSurge: 25%
      maxUnavailable: 25%
    type: RollingUpdate
  template:
    metadata:
      creationTimestamp: null
      labels:
        app: gnlp-0-1-5-3
    spec:
      containers:
      - args:
        - --http2_port=9000
        - --service=gnlp.endpoints.living-bio.cloud.goog
        - --rollout_strategy=managed
        - --backend=grpc://127.0.0.1:80
        - --underscores_in_headers
        image: gcr.io/endpoints-release/endpoints-runtime:1
        imagePullPolicy: IfNotPresent
        name: esp
        ports:
        - containerPort: 9000
          protocol: TCP
...
```

**Comment by qiwzhang** [comment]
Yes, with that flag, ESP nginx will pass client headers with under score to gRPC backend.

**Comment by qiwzhang** [comment]
@lucemia Thanks.

**Comment by qiwzhang** [comment]
Try to clean up the issues... 

Not sure if this feature is still required.  If no, I will close it.

**Comment by ctborg** [comment]
close it out

**Comment by nioann** [comment]
If grpc_set_header is not supported, is there an alternative for passing an x-forwarded-for, or x-real-ip header to the grpc backend?

I'm trying to pass the end user ip to my grpc backend. If there's no solution for this then the ESP can't work as an api endpoint for any use case that requires an ip. In which case, this seems like a very important feature.

The only workaround I can think of is to put another nginx server in front of esp that supports the grpc_set_header and can forward the the correct headers to esp, that will hopefully retain them and send to the grpc backend. That's not a great solution though.

**Comment by ctborg** [comment]
@nioann it looks like just a matter of setting that in the config.

https://github.com/cloudendpoints/esp/blob/aef5aa56006b4e6b70e42d65ad06153cf1f385b1/start_esp/server-auto.conf.template#L77

It looks like there's some code for client_ip_header here:

https://github.com/cloudendpoints/esp/blob/4c86ee209f875a2e123f58459c39035f468822b9/start_esp/start_esp.py#L199

which sets it in the server template here: 
https://github.com/cloudendpoints/esp/blob/4c86ee209f875a2e123f58459c39035f468822b9/start_esp/nginx-auto.conf.template

Try adding `--client_ip_header` to your startup args and see if that works.
for example:
```containers:
- name: esp
  image: gcr.io/endpoints-release/endpoints-runtime:1
  args: [
    "--http_port=8081",
    "--backend=127.0.0.1:8080",
    "--service=SERVICE_NAME",
    "--client_ip_header",
    "--rollout_strategy=managed"
  ]``

**Comment by nioann** [comment]
Thanks @ctborg 
(For posterity) argument above should be --client_ip_header=true

I've tried it but it does not work. My grpc sees the following headers with client_ip_header=true:

[accept-encoding [identity,gzip]],
[:authority [127.0.0.1:52107]],
[x-api-key [<_api_key_>]],
[x-endpoint-api-project-id [<_project_id_>]],
[x-forwarded-user-agent [grpc-go/1.28.0]],
[content-type [application/grpc]],
[user-agent [grpc-c++/1.22.0-dev grpc-c/7.0.0 (linux; chttp2; gale)]],
[grpc-accept-encoding [identity,deflate,gzip]],]

Also from the flag documentation:
"Defines the HTTP header name to extract client IP address"
Seems that the flag is relevant to reading the incoming header to extract the client ip address.

**Comment by ctborg** [comment]
@nioann that makes sense.  Digging in a bit more, it seems your assessment about `grpc_set_header` not being supported is true..

Perhaps fork, and create your image with the required changes. (or send a PR)  e.g. adding `grpc_set_header` argument to [start_esp.py,](https://github.com/cloudendpoints/esp/blob/4c86ee209f875a2e123f58459c39035f468822b9/start_esp/start_esp.py#L151)  then adding line to conditionally add it in the [config](https://github.com/cloudendpoints/esp/blob/4c86ee209f875a2e123f58459c39035f468822b9/start_esp/nginx-auto.conf.template#L235)

e.g.
```
% if location.proto == 'grpc':
      # WARNING: only first backend is used
      grpc_pass ${location.backends[0]} override;      
      % if grpc_set_header
            grpc_set_header X-Real-IP $remote_addr;
      % endif
% else:


```

**Comment by qiwzhang** [comment]
Correct, grpc_set_header is not implemented.  c++ code has to implemented it first. It should go `https://github.com/cloudendpoints/esp/blob/master/src/nginx/grpc.cc#L153`

**Comment by nioann** [comment]
Thanks @ctborg , I might give it a go at some point. For now I've just implemented api key based auth based on api keys on my grpc server, and I've removed ESP from my deployment.

I'm also thinking that if there's another load balancer in front of the end points, which I will probably need for cross-regional LB,  I can populate the headers there.

I don't think this issue should be closed though. Using the user ip on a grpc backend seems like a very common use case.
